azure/services/groups/groups.go

```diff
@@ -35,81 +33,42 @@ const ServiceName = "group"
 // Service provides operations on Azure resources.
 type Service struct {
 	Scope GroupScope
-	aso.Reconciler[ASOType]
+	*aso.Service[ASOType, GroupScope]
 }
 
 // GroupScope defines the scope interface for a group service.
 type GroupScope interface {
 	azure.AsyncStatusUpdater
+	aso.Scope
 	GroupSpec() azure.ASOResourceSpecGetter[*asoresourcesv1.ResourceGroup]
-	GetClient() client.Client
-	ClusterName() string
 }
 
 // New creates a new service.
 func New(scope GroupScope) *Service {
+	svc := aso.NewService[ASOType](ServiceName, scope)
+	svc.Specs = []azure.ASOResourceSpecGetter[ASOType]{scope.GroupSpec()}
+	svc.PostReconcileHook = postReconcileHook
+	svc.PostDeleteHook = postDeleteHook
 	return &Service{
-		Scope:      scope,
-		Reconciler: aso.New[ASOType](scope.GetClient(), scope.ClusterName()),
+		Scope:   scope,
+		Service: svc,
 	}
 }
 
-// Name returns the service name.
-func (s *Service) Name() string {
-	return ServiceName
-}
-
-// Reconcile idempotently creates or updates a resource group.
-func (s *Service) Reconcile(ctx context.Context) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "groups.Service.Reconcile")
-	defer done()
-
-	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureServiceReconcileTimeout)
-	defer cancel()
-
-	groupSpec := s.Scope.GroupSpec()
-	if groupSpec == nil {
-		return nil
-	}
-
-	_, err := s.CreateOrUpdateResource(ctx, groupSpec, ServiceName)
-	s.Scope.UpdatePutStatus(infrav1.ResourceGroupReadyCondition, ServiceName, err)
+func postReconcileHook(scope GroupScope, err error) error {
+	scope.UpdatePutStatus(infrav1.ResourceGroupReadyCondition, ServiceName, err)
```

Comment:
This function is so trivial that it probably doesn't need a test, nonetheless I added one anyway to show how that might be done.

Reply:
+1 on the concept of having a pre/post reconcile hook

That being said, for `UpdatePutStatus` specifically, we do it in all the services so it would be best to include that in the main Service reconcile so it's consistent for every service and we don't have to remember to add it to every single one.

Reply:
+1, will update